Automatically detect and generate Couch reindex migrations

millerdev opened the pull request:
If a PR  changes a Couch design document but does not update the `couchviews.lock` file, the `test-django-migrations.sh` script will fail because `makemigrations` will want to create new a "reindex" migration in the `preindex` app.

:tropical_fish: Review by commit.

## Safety Assurance

### Safety story

Adds automation to ensure a migration is added if a Couch view is changed, which will allow us to safely remove the auto-reindex-on-every-deploy.

### Automated test coverage

The new migration auto-detection patcher has a test to ensure that changes in Django do not disable app-contributed migration detection. Other things added in this PR are exercised by `test-django-migrations.sh`.

### Migrations

The migration in this PR is effectively a no-op. It performs a Couch reindex to ensure that that process is operational from end-to-end.

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Auto-migrate Couch design doc reindex operations

millerdev committed:
Move ReindexCouchViews into corehq.preindex.django_migrations

millerdev committed:
Setup Couch views lock file and reindex migration

The reindex migration in this commit is not strictly needed, but is being created to ensure the process works from end to end. The reindex will run on deploy, but it will be a no-op since no views have changed.

millerdev committed:
Add note about edge case in migrations.lock logic

millerdev committed:
Document run_continuous_indexing_task purpose

joxl commented on corehq/preindex/django_migrations.py:
```diff
@@ -0,0 +1,99 @@
+def write_designs_lock_file(lines, lock_path=LOCK_PATH):
+    """Write a new .lock file."""
+    with open(lock_path, "w") as file:
+        file.write("".join(lines))
```
nit
```suggestion
        file.writelines(lines)
```

millerdev replied:
lol, that copied from here

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/hqscripts/management/commands/makemigrations.py#L108

joxl replied:
:laughing:

joxl commented on corehq/preindex/django_migrations.py:
```diff
@@ -0,0 +1,99 @@
+    if not os.path.exists(lock_path):
+        return True
+    with open(lock_path, "r") as file:
+        return list(file) != lines
```
Is the order of `get_preindex_designs()` deterministic? In other words: if the lines only change in _order_ (if that's even possible), does that justify a "the designs have changed" condition?

millerdev replied:
Good catch!

> Is the order of `get_preindex_designs()` deterministic?

Yes. "Designs" are sorted by design path:

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/corehq/preindex/django_migrations.py#L65

However, Couch design doc manifests apparently were apparently not sorted. Fixed in https://github.com/dimagi/commcare-hq/pull/32544/commits/6d503ca991fd619653212f6053d0a736402d035b

millerdev committed:
Remove ptop_preindex gevent monkey patching

Not needed since manage.py already takes care of that. It was causing
MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors...

millerdev committed:
Patch elastic reindex during test database setup

millerdev committed:
Sort design doc manifest paths

millerdev committed:
writelines

millerdev committed:
Add reindex initial migration dependency

The patch (commit:ff61eca54706fe617cf4c2be7669c7e551844b3a) is still necessary in the test runner because the 0001_bootstrap_es_indexes migration has hard-coded index names, but the ptop_es_manage command operates against test index names, which do not exist at the time that migrations are run.

joxl committed:
Remove `CANONICAL_NAME_INFO_MAP` patch during tests

snopoke reviewed (commented):
I think it would be good to test this on staging prior to merging.

millerdev committed:
Revert "Patch elastic reindex during test database setup"

This reverts commit ff61eca54706fe617cf4c2be7669c7e551844b3a.

millerdev commented:
Rebased to remove unnecessary patch in `HqdbContext`. Thanks for the test fix @joxl

millerdev committed:
Clear elasticsearch and redis on REUSE_DB=reset

millerdev committed:
Move ReindexCouchViews to better place in file

Also rename with a more generic name since it applies to more than Couch views.

millerdev committed:
Run reindex after migrations are applied

millerdev committed:
Update docs about reindex migrations

millerdev committed:
Test reindex on migrate

millerdev commented:
Deployed to staging, and confirmed that the reindex is run at the end of the `migrate` command as expected. Ready for review.

joxl reviewed (approved):
Offlined with @millerdev regarding my curiosity about the decision to make `should_reindex` a module global (this has a smell), which was my only question here. Looks good.
